computer-systems/memory-hierarchy-2-prework/metrics/metrics.go

```diff
@@ -0,0 +1,266 @@
+package metrics
+
+import (
+	"encoding/csv"
+	"log"
+	"math"
+	"os"
+	"strconv"
+	"time"
+)
+
+type UserId int
+//type UserMap map[UserId]*User
+type PaymentsSlice []int
+type AgesSlice []int8
+type PaymentsIndices []int
+
+type Address struct {
+	fullAddress string
+	zip         int
+}
+
+type DollarAmount struct {
+	dollars, cents uint64
+}
+
+type Payment struct {
+	amount DollarAmount
+	time   time.Time
+}
+
+type PaymentsData struct {
+	paymentsCents		PaymentsSlice
+	dollarAmts			[]DollarAmount
+	userIds					[]int
+	times						[]time.Time
+}
+
+
+type UsersData struct {
+	ids						[]int
+	names					[]string
+	ages					AgesSlice
+	addresses			[]Address
+	payments			[]PaymentsIndices
+	paymentsData	PaymentsData
+}
+
+/*
+type User struct {
+	id       UserId
+	name     string
+	age      int
+	address  Address
+	payments []Payment
+}
+*/
+
+func AverageAge(users *UsersData) float64 {
+	
+	ages := users.ages
+	stop := len(ages)
+	stop1 := stop / 4
+	stop2 := stop1 * 2
+	stop3 := stop1 * 3
+
+
+	accChan := make(chan int64, 4)
+	defer close(accChan)
+
+	go procAccAge(0, stop1, ages, accChan)
+	go procAccAge(stop1, stop2, ages, accChan)
+	go procAccAge(stop2, stop3, ages, accChan)
+	go procAccAge(stop3, stop, ages, accChan)
+
+	var accSum int64 = 0
+	count := 0
+	for acc := range accChan {
+		accSum += acc
+		count += 1
+		if count == 4 {
```

Comment:
https://gobyexample.com/range-over-channels fwiw I was doing this break too but then I saw you can close a channel and ranging over a closed channel stops once it's empty

Reply:
Oh I see. I knew you could could stop when closed, but I didn't think about closing the channel being currently looped over. Thanks!

Before I had a separate go routine wait and close the channel, but I think I set it up wrong and it ended up with some race conditions.